File: Src/Engine/Editor/SceneGraph/ScenesGraphNode.cpp

#include "ScenesGraphNode.h"

#include "Editor/EditorApp.h"
#include "Editor/Modules/SceneGraphModule.h"

namespace SE::Editor
{
	ScenesGraphNode* ScenesGraphNode::__GetParentNode()
	{
		return parentNode;
	}

	void ScenesGraphNode::__SetParentNode(ScenesGraphNode* value)
	{
		if (parentNode != value)
		{
			if (parentNode != nullptr)
			{
				parentNode->ChildNodes.Remove(this);
			}
			parentNode = value;
			if (parentNode != nullptr)
			{
				parentNode->ChildNodes.Add(this);
			}

			OnParentChanged();
		}
	}

	RootGraphNode* ScenesGraphNode::__GetRoot()
	{
		if (ParentNode == nullptr)
		{
			return nullptr;
		}

		return ParentNode->Root;
	}

	void ScenesGraphNode::Dispose()
	{
		OnDispose();

		// Unlink from the parent
		if (parentNode != nullptr)
		{
			parentNode->ChildNodes.Remove(this);
			parentNode = nullptr;
		}
	}

	void ScenesGraphNode::OnDispose()
	{
		// Call deeper
		for (int i = 0; i < ChildNodes.Count(); i++)
		{
			ChildNodes[i]->OnDispose();
		}

		EditorApp::Ins().sceneGraphModule->Nodes.Remove(ID);
	}


	ScenesGraphNode::ScenesGraphNode(UID id)
	{
		m_ID = id;
		ScenesGraphNode* duplicate;
		SceneGraphModule* sceneGraphModule = EditorApp::Ins().sceneGraphModule;;
		if (sceneGraphModule->Nodes.TryGet(id, duplicate) && duplicate != nullptr)
		{
			LOG_WARNING("Scene", "Duplicated Scene Graph node with ID {0} of type '{1}'", id, duplicate->GetTypeInfo()->friendlyName);
		}
		sceneGraphModule->Nodes[id] = this;
	}

	::SE::Transform ScenesGraphNode::__GetTransform()
	{
		return m_Transform;
	}

	void ScenesGraphNode::__SetTransform(::SE::Transform value)
	{
		m_Transform = value;
	}

} // SE

File: Src/Engine/Editor/SceneGraph/ScenesGraphNode.cs
using System;
using System.Collections.Generic;

namespace SE.Editor.SceneGraph
{
    /// <summary>
    /// Base node for the editor scene graph. Runtime objects stay native; this
    /// type owns only editor hierarchy state and its managed lifetime.
    /// </summary>
    public abstract class ScenesGraphNode : IDisposable
    {
        private ScenesGraphNode? m_ParentNode;
        private SE.Transform m_Transform = SE.Transform.Identity;
        private int m_OrderInParent;
        private bool m_IsDisposed;

        protected ScenesGraphNode(SceneGraphFactory factory, Guid id)
        {
            Factory = factory ?? throw new ArgumentNullException(nameof(factory));
            ID = id;
            Factory.RegisterNode(this);
        }

        internal SceneGraphFactory Factory { get; }

        /// <summary>
        /// Children in scene hierarchy order.
        /// </summary>
        public List<ScenesGraphNode> ChildNodes { get; } = new();

        public abstract string Name { get; }
        public Guid ID { get; }
        public virtual RootGraphNode? Root => ParentNode?.Root;
        public virtual Transform Transform
        {
            get => m_Transform;
            set => m_Transform = value;
        }
        public abstract SceneGraphNode? ParentScene { get; }
        public virtual int OrderInParent
        {
            get => m_OrderInParent;
            set => m_OrderInParent = value;
        }
        public bool IsDisposed => m_IsDisposed;

        public virtual ScenesGraphNode? ParentNode
        {
            get => m_ParentNode;
            set
            {
                if (ReferenceEquals(m_ParentNode, value))
                    return;
                if (ReferenceEquals(this, value))
                    throw new InvalidOperationException("A scene graph node cannot parent itself.");

                m_ParentNode?.ChildNodes.Remove(this);
                m_ParentNode = value;
                if (m_ParentNode != null && !m_ParentNode.ChildNodes.Contains(this))
                    m_ParentNode.ChildNodes.Add(this);

                OnParentChanged();
            }
        }

        public abstract SE.Object? GetEditableObject();

        /// <summary>
        /// Releases this node tree and unlinks this node from its parent.
        /// </summary>
        public virtual void Dispose()
        {
            if (m_IsDisposed)
                return;

            OnDispose();
            if (m_ParentNode != null)
            {
                m_ParentNode.ChildNodes.Remove(this);
                m_ParentNode = null;
            }
        }

        /// <summary>
        /// Releases descendants and removes this node from the graph registry.
        /// </summary>
        public virtual void OnDispose()
        {
            if (m_IsDisposed)
                return;

            m_IsDisposed = true;
            ScenesGraphNode[] children = ChildNodes.ToArray();
            foreach (ScenesGraphNode child in children)
                child.OnDispose();
            ChildNodes.Clear();

            Factory.UnregisterNode(this);
        }

        protected virtual void OnParentChanged()
        {
        }
    }
}
